### PR TITLE
Replace CGLib with ByteBuddy

### LINKED ISSUE #577: Random error 1.13.2 using Latest build on #425
It seems to throw this error on boot of the server.
```

WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.comphenix.net.sf.cglib.core.ReflectUtils$1 (file:/mnt/raid0/AMP/.ampdata/instances/ltcraft-main/Minecraft/plugins/ProtocolLib%20(2).jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.comphenix.net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```
Below is a paste bin of my last log:

[](https://pastebin.com/LvVDw46d)

Let me know if i can provide any other details.

### LINKED ISSUE #603: Illegal reflective access with spigot 1.14
_Follow this template except for feature requests. Use pastebin when providing /protocol dump and any relevant errors._

Make sure you've done the following:
- [x] You're using the latest build for your server version
- [x] This isn't an issue caused by another plugin
- [x] You've checked for duplicate issues
- [x] You didn't use `/reload`

Debug paste link:

``WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.comphenix.net.sf.cglib.core.ReflectUtils$1 (file:/home/ieti/minecraft/plugins/ProtocolLib.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.comphenix.net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
``

Description and relevant errors:

This warning appears with latest 4.5 snapshot from gradle jenkins.

### LINKED ISSUE #886: [1.16.1] An illegal reflective access operation has occurred
**Describe the bug**

After the server is done loading it spits out a an illegal reflective access warning msg. 
stop and start again, it doesn't seem to show it.

```
[04:47:56] [Server thread/INFO]: Done (27.375s)! For help, type "help"
>WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.comphenix.net.sf.cglib.core.ReflectUtils$1 (file:/Users/floris/MinecraftServer/z-serverlistplus/plugins/ProtocolLib.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.comphenix.net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

>
```

**To Reproduce**
load server, go to console, watch it show up.

**Expected behavior**
did not expect it to show up at all.

**Screenshots**

**Version Info**
Provide your ProtocolLib install info with `/protocol dump` through pastebin.
[dump-2020-06-29_04.53.11.txt](https://github.com/dmulloy2/ProtocolLib/files/4843503/dump-2020-06-29_04.53.11.txt)

**Additional context**
```
>ver
[04:54:03] [Server thread/INFO]: This server is running CraftBukkit version git-Spigot-ad703da-e2403a3 (MC: 1.16.1) (Implementing API version 1.16.1-R0.1-SNAPSHOT)
[04:54:03] [Server thread/INFO]: You are running the latest version
>ver ProtocolLib
[04:54:06] [Server thread/INFO]: ProtocolLib version 4.6.0-SNAPSHOT-b466
[04:54:06] [Server thread/INFO]: Provides read/write access to the Minecraft protocol.
[04:54:06] [Server thread/INFO]: Authors: dmulloy2 and comphenix
>ver ServerListPlus
[04:54:09] [Server thread/INFO]: ServerListPlus version 3.5.0-SNAPSHOT
[04:54:09] [Server thread/INFO]: A flexible Minecraft plugin to customize the appearance of your server in the server list
[04:54:09] [Server thread/INFO]: Website: https://git.io/slp
[04:54:09] [Server thread/INFO]: Author: Minecrell (https://github.com/Minecrell)
>pl
[04:54:18] [Server thread/INFO]: Plugins (5): Vault, AntiFire, ProtocolLib, ServerListPlus, CMI
>
```

### LINKED ISSUE #947: Future proofing ProtocolLib for Java 11+ (replacing cglib with ByteBuddy?)
I've recently noticed several bug reports regarding `Illegal reflective access` in com.comphenix.net.sf.cglib.core.ReflectUtils triggered through ProtocolLib on various sites ([[1]](https://github.com/dmulloy2/ProtocolLib/issues/886), [[2]](https://github.com/MilkBowl/VaultAPI/issues/91), [[3]](https://www.spigotmc.org/threads/protocollib.36230/page-52#post-3270342), [[4]](https://github.com/aadnk/ProtocolLib/issues/160)). While this is currently just a warning, this has the potential to become a bigger problem in the future. Quoting the warning message:
> `WARNING: All illegal access operations will be denied in a future release`

This is due to stricter reflection rules in Java 9 and above, and I'm not sure to what extent this limits ProtocolLib's functionality today. A [comment](https://github.com/dmulloy2/ProtocolLib/issues/603#issuecomment-490207994) from the repo owner on a related issue from May 2019 suggests that this will not be worked on before Minecraft switches to a newer java version.

Nonetheless, we can already see a considerably large increase in Java 9+ adoption for Minecraft in the past 6 months, so this is bound to become a problem in the (near) future:

| Date     | Java 8 | Java 9+   | Reference                                                                                 |
|----------|--------|-----------|-------------------------------------------------------------------------------------------|
| Feb 2020 | 96.8%  | 1.0%      | [bStats.org](https://web.archive.org/web/20200201071929/https://bstats.org/global/bukkit) |
| Aug 2020 | 84.7%  | **15.3%** | [bStats.org](https://web.archive.org/web/20200823123415/https://bstats.org/global/bukkit) |

#### The actual bug (cglib)

The actual bug is in a dependency of ProtocolLib, specifically in [cglib](https://github.com/cglib/cglib), which is just shaded into `com.comphenix.net.sf.cglib`.
I've found this related open [issue](https://github.com/cglib/cglib/issues/133) in cglib from August 2018, as well as [a second](https://github.com/cglib/cglib/issues/175) and [a third](https://github.com/cglib/cglib/issues/176) one. The problem is now known for over two years, but nothing has happened.

Apparently there is no ongoing development to cglib ([last commit](https://github.com/cglib/cglib/commit/975e481faf39c91b8ac5b9b3d62822b7c52c5f47) was **13 months ago**), which is concerning. On multiple occasions I've read the suggestion to switch to Byte Buddy instead for newer java versions, so maybe @dmulloy2 might consider doing that.

Anyway, just wanted to raise awareness of this issue, before it becomes an actual problem and a solution might be rushed.

### LINKED ISSUE #986: An illegal reflective access operation has occurred
**Describe the bug**
Got a warning in my server console sometime after a restart, only plugin that I run that uses ProtocolLib is SuperVanish as far as I know.
Not sure how much of an issue this is, doesn't seem to affect the server negatively atm.

**The warning:**
>WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.comphenix.net.sf.cglib.core.ReflectUtils$1 (file:/home/container/plugins/ProtocolLib-4.6.0-SNAPSHOT.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.comphenix.net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

**To Reproduce**
Warning did appear after a restart but I don't know how much time it took after the restart as the console didn't report a time.
No players were online at that moment.

**Version Info**
https://pastebin.com/tAvjfkir

### LINKED ISSUE #1029: Found this warning.
**Describe the bug**
`
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.comphenix.net.sf.cglib.core.ReflectUtils$1 (file:/home/network/paintball/plugins/ProtocolLib.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.comphenix.net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
`

**Version Info**
https://gist.github.com/MithrandirCraft/9fe0b1c9f9292a7dbd81a5cdc89e248a

### PR DESCRIPTION
**Pull request by PimvanderLoos**
This PR replaces CGLib with ByteBuddy. This closes issues #577, #603, #886, and #947.

I have marked this is a draft because there are a few items I am unsure about.
- What is the desired way of dealing with exceptions? Specifically [here](https://github.com/PimvanderLoos/ProtocolLib/blob/e704336ee12820323136fe5d41f0e7c55dfc6c17/src/main/java/com/comphenix/protocol/utility/MinecraftMethods.java#L208), [here](https://github.com/PimvanderLoos/ProtocolLib/blob/e704336ee12820323136fe5d41f0e7c55dfc6c17/src/main/java/com/comphenix/protocol/events/SerializedOfflinePlayer.java#L286), and [here](https://github.com/PimvanderLoos/ProtocolLib/blob/e704336ee12820323136fe5d41f0e7c55dfc6c17/src/main/java/com/comphenix/protocol/injector/server/TemporaryPlayerFactory.java#L174).
- What would be the desired way of testing [TileEntityAccessor$findMethodUsingByteBuddy(T)](https://github.com/PimvanderLoos/ProtocolLib/blob/bytebuddy/src/main/java/com/comphenix/protocol/wrappers/nbt/TileEntityAccessor.java#L178). I've written tests for the other bigger changes, but this one required the server and it looks like the integration tests are currently disabled, so I was unsure how to proceed.

To properly link the issues (I can't select them manually for some reason): Closes #577, closes #603, closes #886, closes #947, closes #986, closes #1029

### CODE REVIEW

**Comment by dmulloy2** [comment]
Seems like a relatively straightforward translation from CGLib to ByteBuddy.

I think generally the way to test the methods would be to make sure they get found. If you want to go above and beyond on that, maybe trying an invocation on a mocked tile could work.

I'm assuming the primary benefit here is that ByteBuddy is actively maintained and works with Java 9+. Are there any benefits besides that? I would be interested in the performance difference between the two if that wouldn't be too hard to get.

In any case, thank you for all the work!

**Comment by PimvanderLoos** [comment]
All right, I'll figure something out for the test.

Those are indeed the main benefits of ByteBuddy over CGLib. ByteBuddy also [appears to be slightly faster](https://bytebuddy.net/#/tutorial) in certain situations. However, when I tried to run a benchmark to see if that's the case in this situation with this implementation, I found out my translation is a little too literal. The generated classes work, but only once (which is why the test didn't catch it). After that, it'll just throw an error, as the class has already been loaded.

I'll work on an updated version and some benchmarks soon.

File: src/main/java/com/comphenix/protocol/events/SerializedOfflinePlayer.java [hunk 1]
```diff
@@ -257,28 +258,35 @@ public Player getProxyPlayer() {
...
+						// Invoke our on method
+						return offlineMethod.invoke(SerializedOfflinePlayer.this, args);
+					}))
+					.make()
+					.load(ByteBuddyFactory.getInstance().getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+					.getLoaded()
+					.getDeclaredConstructor()
+					.newInstance();
+
+			// TODO:P What's the desired way to deal with these?
```
[the remark this answers is not in the document]
Comment (PimvanderLoos) [reply src/main/java/com/comphenix/protocol/events/SerializedOfflinePlayer.java:286]: Thank you for the clarification. I'll update the code to do this.

**Comment by PimvanderLoos** [comment]
I've updated all the byte buddy stuff and I've run some benchmarks. The [results](https://raw.githubusercontent.com/PimvanderLoos/ProtocolLib-BenchMark/master/results/results.txt) and the methodology of the benchmarks can be found [here](https://github.com/PimvanderLoos/ProtocolLib-BenchMark). 

To quickly summarize the results, ByteBuddy appears to be quite a bit faster across the board. The only exception is in the situation of accessing any of the online methods of a Proxy Player, which throws an UnsupportedOperationException. However, it is still quite close and I'd argue that it's probably better to check if the player object is generated or not (using instanceof ByteBuddyGenerated) before accessing the methods rather than relying on exception handling as this will be much faster regardless of the implementation. In a quick benchmark that's not listed in that repo, an instanceof check was about 3603 times faster than throwing/catching an exception.



However, I ran into some issues with the TileEntityAccessor class. 
Because of [this](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/19507baf8b7903427bc3efab7118de6e7c1c931e#src/main/java/org/bukkit/craftbukkit/block/CraftSign.java) commit to CraftBukkit (pushed right after 1.12.1), subclasses of BlockState no longer have a TileEntity field. Instead, they extend CraftBlockEntityState which does have this field. Because of this, it cannot find the TileEntity field [here](https://github.com/dmulloy2/ProtocolLib/blob/bdaa843f2dd8f05fdc48a30786ecd1a23416c652/src/main/java/com/comphenix/protocol/wrappers/nbt/TileEntityAccessor.java#L271), causing the TileEntityAccessor code not to be run at all.

I ended up running the test manually on 1.11. However, there was another issue where the result of [this](https://github.com/dmulloy2/ProtocolLib/blob/bdaa843f2dd8f05fdc48a30786ecd1a23416c652/src/main/java/com/comphenix/protocol/wrappers/nbt/TileEntityAccessor.java#L198) method call didn't include the methods it was supposed to be looking for. After hacking the code a bit so it does include the correct methods, I at least found that the ByteBuddy-related part worked as intended.

Considering how long this issue has been here and that the only places where the TileEntityAccess class is used are [here](https://github.com/dmulloy2/ProtocolLib/blob/bdaa843f2dd8f05fdc48a30786ecd1a23416c652/src/main/java/com/comphenix/protocol/wrappers/nbt/NbtFactory.java#L256) and [here](https://github.com/dmulloy2/ProtocolLib/blob/bdaa843f2dd8f05fdc48a30786ecd1a23416c652/src/main/java/com/comphenix/protocol/wrappers/nbt/NbtFactory.java#L242), neither of which is used anywhere else as far as I could see, I wonder if it could not be removed altogether?

**Comment by dmulloy2** [comment]
Great work! The speed increase is definitely a plus.

I was under the impression that plugins still used those methods from NBTFactory. Unless you're saying they don't currently work?

**Comment by PimvanderLoos** [comment]
Well, they don't work at all on 1.12.2+ (also documented in [this](https://github.com/aadnk/ProtocolLib/issues/143) issue), and on 1.11 it works as long as it can retrieve the read/write compounds using ASM. Though the methods it returns aren't quite the ones I'd expect. On MC 1.11 and the PL master branch, the read method found by ASM is `void a(NBTTagCompound)` and the write method is `NBTTagCompound c(NBTTagCompound)`, which is the opposite of those found [here](https://github.com/dmulloy2/ProtocolLib/blob/bdaa843f2dd8f05fdc48a30786ecd1a23416c652/src/main/java/com/comphenix/protocol/wrappers/nbt/TileEntityAccessor.java#L77-L91) when changing the NbtFactory such that it retrieves the `TileEntity` field from the BlockState's superclass.

**Comment by dmulloy2** [comment]
@PimvanderLoos do you think this PR is ready? (it's still marked as a draft)

**Comment by PimvanderLoos** [comment]
Yes, I think it's ready. I've removed the draft flag. 

To properly link the issues (I can't select them manually for some reason): Closes #577, closes #603, closes #886, closes #947, closes #986

**Pull request merged**